ui/store/action-queue/index.js

```diff
@@ -0,0 +1,200 @@
+import pify from 'pify';
+import { isManifestV3 } from '../../../shared/modules/mv3.utils';
+
+// // A simplified pify maybe?
+// function pify(apiObject) {
+//   return Object.keys(apiObject).reduce((promisifiedAPI, key) => {
+//     if (apiObject[key].apply) { // depending on our browser support we might use a nicer check for functions here
+//       promisifiedAPI[key] = function (...args) {
+//         return new Promise((resolve, reject) => {
+//           return apiObject[key](
+//             ...args,
+//             (err, result) => {
+//               if (err) {
+//                 reject(err);
+//               } else {
+//                 resolve(result);
+//               }
+//             },
+//           );
+//         });
+//       };
+//     }
+//     return promisifiedAPI;
+//   }, {});
+// }
+
+let background = null;
+let promisifiedBackground = null;
+
+const actionRetryQueue = [];
+
+function failQueue() {
+  actionRetryQueue.forEach(({ reject }) =>
+    reject(
+      Error('Background operation cancelled while waiting for connection.'),
+    ),
+  );
+}
+
+/**
+ * Drops the entire actions queue. Rejects all actions in the queue unless silently==true
+ * Does not affect the single action that is currently being processed.
+ *
+ * @param {boolean} [silently]
+ */
+export function dropQueue(silently) {
+  if (!silently) {
+    failQueue();
+  }
+  actionRetryQueue.length = 0;
+}
+
+// add action to queue
+const executeActionOrAddtoRetryQueue = (item) => {
+  if (actionRetryQueue.some((act) => act.actionId === item.actionId)) {
+    return;
+  }
+
+  if (background.connectionStream.readable) {
+    executeAction({
+      action: item,
+      disconnectSideeffect: () => actionRetryQueue.push(item),
+    });
+  } else {
+    actionRetryQueue.push(item);
+  }
+};
+
+/**
+ * Promise-style call to background method
+ * In MV2: invokes promisifiedBackground method directly.
+ * In MV3: action is added to retry queue, along with resolve handler to be executed on completion,
+ *  the queue is then immediately processed if background connection is available.
+ *  On completion (successful or error) the action is removed from the retry queue.
+ *
+ * @param {string} method - name of the background method
+ * @param {Array} [args] - arguments to that method, if any
+ * @param {any} [actionId] - if an action with the === same id is submitted, it'll be ignored if already in queue waiting for a retry.
+ * @returns {Promise}
+ */
+export function submitRequestToBackground(
+  method,
+  args = [],
+  actionId = Date.now() + Math.random(), // current date is not guaranteed to be unique
```

Comment:
Is this `actionId` param being used anywhere? Or can we move this random generator into the function itself?

Reply:
yes following PRs are using it, plus this flexibility for functions to be able to provide their own actionId is useful